### PR TITLE
support  agent image config

### PR DESCRIPTION
**Pull request by 804e**
## What type of PR is this?

- support  "agent.image" configuration

#### Issue(s) number: 

[#209](https://github.com/vesoft-inc/nebula-operator/issues/209)

### CODE REVIEW

**Comment by wey-gu** [comment]
@804e thanks for the contribution! Could you please sign the above CLA?

@MegaByte875 could you please take a look at this change?

**Comment by 804e** [comment]
have logged in CLA @wey-gu

**Comment by wey-gu** [comment]
> have logged in CLA @wey-gu

Could you please ensure the github account(that you signed CLA) matches the commit auther's email? For now it seems not the case.

**Comment by 804e** [comment]
i've logged in and recheck CLA,now it's "All checks have passed"

**Comment by MegaByte875** [comment]
@804e Thanks for you PR

File: apis/apps/v1alpha1/nebulacluster_types.go [hunk 1]
```diff
@@ -219,6 +221,11 @@ type ExporterSpec struct {
 	MaxRequests int32 `json:"maxRequests,omitempty"`
 }
 
+// AgentSpec defines the desired state of Agent
+type AgentSpec struct {
```
Comment (MegaByte875) [inline apis/apps/v1alpha1/nebulacluster_types.go:225]: There is no need to embed PodSpec here as it is only used for a component. Instead, you can specify the necessary fields within the AgentSpec, which are three in total.
Image
Version
Resources

By the way, I think renaming AgentSpec to AgentContainerSpec would be better.

Comment (804e) [reply apis/apps/v1alpha1/nebulacluster_types.go]: Thank you for your suggestion. I have revised it

File: apis/apps/v1alpha1/nebulacluster_common.go [hunk 2]
```diff
@@ -214,10 +214,12 @@ func generateAgentContainer(c NebulaClusterComponentter, init bool) corev1.Conta
...
 
+	agentImage := getImage(c.GetNebulaCluster().Spec.Agent.Image, c.GetNebulaCluster().Spec.Agent.Version, defaultMetadImage)
 	container := corev1.Container{
-		Name:    AgentSidecarContainerName,
-		Image:   defaultAgentImage,
-		Command: agentCmd,
+		Name:      AgentSidecarContainerName,
+		Image:     agentImage,
+		Command:   agentCmd,
+		Resources: c.GetNebulaCluster().Spec.Agent.Resources,
```
Comment (MegaByte875) [inline apis/apps/v1alpha1/nebulacluster_common.go:222]: Please rebase to master branch, here has a commit conflict.

Comment (804e) [reply apis/apps/v1alpha1/nebulacluster_common.go:232]: have already rebase

Comment (MegaByte875) [reply apis/apps/v1alpha1/nebulacluster_common.go:232]: The function 'getImage' has been removed. And I see you merged my previous PR.

Comment (804e) [reply apis/apps/v1alpha1/nebulacluster_common.go]: i've delete getImage function call,but i got some error on check
```
[root@base01 nebula-operator]# make check
go mod tidy
# go fmt ./...
/opt/goworkspace/bin/gofumpt -w -l ./
go vet ./...
# github.com/vesoft-inc/nebula-operator/pkg/ngctl/cmd/info
pkg/ngctl/cmd/info/clusterinfo.go:97:39: nc.GraphdComponent().GetImage undefined (type "github.com/vesoft-inc/nebula-operator/apis/apps/v1alpha1".NebulaClusterComponent has no field or method GetImage)
pkg/ngctl/cmd/info/clusterinfo.go:246:65: undefined: v1alpha1.NebulaClusterComponentter
# github.com/vesoft-inc/nebula-operator/pkg/scheduler/extender/predicates
pkg/scheduler/extender/predicates/topology.go:85:24: component.GetReplicas undefined (type v1alpha1.NebulaClusterComponent has no field or method GetReplicas)
pkg/scheduler/extender/predicates/topology.go:125:50: undefined: v1alpha1.NebulaClusterComponentter
# github.com/vesoft-inc/nebula-operator/pkg/controller/nebulacluster
vet: pkg/controller/nebulacluster/nebula_cluster_condition_updater_test.go:78:25: PodSpec not declared by package v1alpha1
# github.com/vesoft-inc/nebula-operator/tests/e2e/nebulacluster
tests/e2e/nebulacluster/util.go:60:5: unknown field 'PodSpec' in struct literal of type "github.com/vesoft-inc/nebula-operator/apis/apps/v1alpha1".GraphdSpec
tests/e2e/nebulacluster/util.go:60:23: undefined: v1alpha1.PodSpec
tests/e2e/nebulacluster/util.go:75:5: unknown field 'PodSpec' in struct literal of type "github.com/vesoft-inc/nebula-operator/apis/apps/v1alpha1".MetadSpec
tests/e2e/nebulacluster/util.go:75:23: undefined: v1alpha1.PodSpec
tests/e2e/nebulacluster/util.go:98:5: unknown field 'PodSpec' in struct literal of type "github.com/vesoft-inc/nebula-operator/apis/apps/v1alpha1".StoragedSpec
tests/e2e/nebulacluster/util.go:98:23: undefined: v1alpha1.PodSpec
make: *** [vet] Error 1
```

Comment (804e) [reply apis/apps/v1alpha1/nebulacluster_common.go]: my fault, ignore it

Comment (804e) [reply apis/apps/v1alpha1/nebulacluster_common.go]: > i've delete getImage function call,but i got some error on check
> 
> ```
> [root@base01 nebula-operator]# make check
> go mod tidy
> # go fmt ./...
> /opt/goworkspace/bin/gofumpt -w -l ./
> go vet ./...
> # github.com/vesoft-inc/nebula-operator/pkg/ngctl/cmd/info
> pkg/ngctl/cmd/info/clusterinfo.go:97:39: nc.GraphdComponent().GetImage undefined (type "github.com/vesoft-inc/nebula-operator/apis/apps/v1alpha1".NebulaClusterComponent has no field or method GetImage)
> pkg/ngctl/cmd/info/clusterinfo.go:246:65: undefined: v1alpha1.NebulaClusterComponentter
> # github.com/vesoft-inc/nebula-operator/pkg/scheduler/extender/predicates
> pkg/scheduler/extender/predicates/topology.go:85:24: component.GetReplicas undefined (type v1alpha1.NebulaClusterComponent has no field or method GetReplicas)
> pkg/scheduler/extender/predicates/topology.go:125:50: undefined: v1alpha1.NebulaClusterComponentter
> # github.com/vesoft-inc/nebula-operator/pkg/controller/nebulacluster
> vet: pkg/controller/nebulacluster/nebula_cluster_condition_updater_test.go:78:25: PodSpec not declared by package v1alpha1
> # github.com/vesoft-inc/nebula-operator/tests/e2e/nebulacluster
> tests/e2e/nebulacluster/util.go:60:5: unknown field 'PodSpec' in struct literal of type "github.com/vesoft-inc/nebula-operator/apis/apps/v1alpha1".GraphdSpec
> tests/e2e/nebulacluster/util.go:60:23: undefined: v1alpha1.PodSpec
> tests/e2e/nebulacluster/util.go:75:5: unknown field 'PodSpec' in struct literal of type "github.com/vesoft-inc/nebula-operator/apis/apps/v1alpha1".MetadSpec
> tests/e2e/nebulacluster/util.go:75:23: undefined: v1alpha1.PodSpec
> tests/e2e/nebulacluster/util.go:98:5: unknown field 'PodSpec' in struct literal of type "github.com/vesoft-inc/nebula-operator/apis/apps/v1alpha1".StoragedSpec
> tests/e2e/nebulacluster/util.go:98:23: undefined: v1alpha1.PodSpec
> make: *** [vet] Error 1
> ```

my fault, ignore it

**Comment by MegaByte875** [comment]
@804e Your code looks good to me, please squash multiple commits into one commit and your branch outdated so need to be rebased on the master banch .

**Comment by wey-gu** [comment]
@804e could you please help do something like this?:

- create a branch of the latest master
- do changes to this PR to this branch with one or more commits
- force push it to your branch(804e:master)

Thanks!

**Comment by 804e** [comment]
already merge squash as a commit @wey-gu @MegaByte875

**Review by wey-gu — APPROVED** [review]

**Review by MegaByte875 — APPROVED** [review]
LGTM

**Comment by wey-gu** [comment]
Thanks a lot @804e and @MegaByte875 for the fantastic contribution and collaboration!

cc @abby-cyber for a documentation-related change!

**Pull request merged**